src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs

```diff
@@ -0,0 +1,489 @@
+            /// <param name="count">The number of characters to write from the character array.</param>
+            /// <returns>A task representing the asynchronous write operation.</returns>
+            public override async Task WriteAsync(char[] value, int index, int count)
+            {
+                if (this.isWritingJson)
+                {
+                    ReadOnlyMemory<char> charValuesToWrite = value.AsMemory().Slice(index, count);
+                    await this.WritingCharValuesInChunksAsync(charValuesToWrite).ConfigureAwait(false);
+                }
+                else
+                {
+                    ReadOnlyMemory<char> charValuesToWrite = value.AsMemory().Slice(index, count);
+                    await this.WritingCharValuesInChunksWithoutEscapingAsync(charValuesToWrite).ConfigureAwait(false);
+                }
+            }
+
+            /// <summary>
+            /// Writes the characters represented by the provided read-only span in chunks, processing them for escaping if necessary, and flushing the writer if the buffer threshold is reached.
+            /// </summary>
+            /// <param name="value">The read-only span containing the characters to be written.</param>
+            private void WritingCharValuesInChunks(ReadOnlySpan<char> value)
+            {
+                // Process the input value in chunks
+                for (int i = 0; i < value.Length; i += ODataUtf8JsonWriter.chunkSize)
+                {
+                    int remainingChars = Math.Min(ODataUtf8JsonWriter.chunkSize, value.Length - i);
+                    bool isFinalBlock = false;
+                    ReadOnlySpan<char> chunk = value.Slice(i, remainingChars);
+
+                    int firstIndexToEscape = this.jsonWriter.NeedsEscaping(chunk);
+
+                    // If the buffer is not empty, then we copy the chars from the buffer
+                    // to the current chunk being processed.
+                    if (numOfCharsNotWrittenFromPreviousChunk > 0)
+                    {
+                        // Get unprocessed chars from the buffer.
+                        ReadOnlySpan<char> charsNotProcessedFromPreviousChunk = this.buffer.AsSpan().Slice(0, numOfCharsNotWrittenFromPreviousChunk);
+                        int totalLength = charsNotProcessedFromPreviousChunk.Length + chunk.Length;
+
+                        char[] combinedArray = ArrayPool<char>.Shared.Rent(totalLength);
+
+                        // Copy chars from charsNotProcessedFromPreviousChunk to the combined array
+                        charsNotProcessedFromPreviousChunk.CopyTo(combinedArray);
+
+                        // Copy chars from chunk to the combined array, starting from the end of charsNotProcessedFromPreviousChunk
+                        chunk.CopyTo(combinedArray.AsSpan().Slice(charsNotProcessedFromPreviousChunk.Length));
+

```

Comment:
REF: https://github.com/OData/odata.net/pull/2880/files#r1551128563
Maybe this is where we should have:
```
int firstIndexToEscape = this.jsonWriter.NeedsEscaping(charsNotProcessedFromPreviousChunk);
if (firstIndexToEscape == <WHATEVER VALUE IS RETURNED IF SUCH A CHAR DOESN'T EXIST IN THE CHUCK>)
{
    firstIndexToEscape = this.jsonWriter.NeedsEscaping(chunk) + charsNotProcessedFromPreviousChunk.Length;
}
```
Otherwise the current math is wrong since the `firstIndexToEscape` doesn't account for chars not processed from previous chunk...

Reply:
We also need to get the firstIndexToEscape in cases where we do not have unprocessed bytes.

Reply:
So this method should like this: 
```c#
if (this.numOfCharsNotWrittenFromPreviousChunk > 0)
{
     // here we get the combinedArray that has the chunk and the unprocessed bytes.. 
      // Then the combined array is what we pass to the NeedEscaping method like: 

int firstIndexToEscape = this.jsonWriter.NeedsEscaping(combinedArray);
}else
{
int firstIndexToEscape = this.jsonWriter.NeedsEscaping(chunk);
}
```

Reply:
Although we have a lot of tests on this. I tend to believe it's hard to have unprocessed chars that require escaping. Just to be safe.. I'll make the above change.